### ISSUE TITLE
Problem when passing functions to cells

### ISSUE THREAD

**Issue by draben** [juba/robservable#38]
Within Observable, suppose we have a cell that is a simple function, perhaps for formatting to dollars in $M:

<code>format = d => \`$${d}M\`</code>

From robservable, I expected to be able to modify this cell via an input as follows:

<code>robservable(..., input = list(format = htmlwidgets::JS("d =>\`${d}\`"))</code>

However, it appears that the function is treated as a string literal and therefore the notebook throws a "format is not a function" error.  Assuming that updating cells with functions is supported, what is the appropriate syntax?

Thanks in advance!

**Comment by juba** [comment]
Sorry for the delay of this answer. I'm not surprised this doesn't work, I think that the Observable runtime only allows to redefine cell as values, not as code : 

https://github.com/observablehq/runtime

But I may be wrong and I'll try to take a deeper look soon.

**Comment by draben** [comment]
Interesting.  That would seem like a large missing piece of functionality in the runtime, but I admit to not understanding everything at the link you provided.  In reading through the variables section, I wonder if there are ideas there.  Specifically, the part that begins "To define an anonymous variable b that takes foo as input:" suggests function cells can be defined.  Maybe robservable could treat the input list argument as is and support another function list argument differently?  No clue.

As with the last issue, I very much appreciate your willingness to support this package.  Thank you!

**Comment by juba** [comment]
So, after a bit of digging, it appears that it is possible right out of the box with current `robservable`, but you have to write it like this : 

```r
input = list(format  = htmlwidgets::JS("() => (d => (`${d} dollars`))"))
```
The reason that if you wanted to do it in pure JavaScript with the Observable runtime API, you would use something like : 

```js
module.redefine("format", [], () => ((d) => `${d} dollars`))
```

ie you would redefine `format` by passing a function without argument which returns the new function you want to assign to the cell. 

Note that this will work only if your function is "pure" ie doesn't depend from the values of another cell. If your function was `format = (y) => (cell1 + y)` where `cell1` is the value of another cell, you would do it in JavaScript with : 

```js
module.redefine("format", ["cell1"], (cell) => ((y) => cell * y))
```
but it isn't supported in `robservable` for now.

**Comment by timelyportfolio** [comment]
@juba You beat me to it.  I had planned to play with this today.  Let me know if you think we should explore adding support for this.

**Comment by juba** [comment]
Thanks for planning to take a look at it !
Adding support for this should not be that hard, I think we just have to give a way to users to provide both a `JS` string and an array of argument names in `input`, and then to modify the call to `redefine` if such parameters are passed. If you want to implement it, that could only be useful, even if I don't think this will be a very frequent use case.

**Comment by draben** [comment]
Thanks to you both!  My specific issue has been addressed (just verified), so let me know if you want me to close the issue.  That said, keeping it open to track code improvements might be nice.

On frequency of use, consider how excellent Observable can be to create visualization templates for data science teams.  The format function example above might be too simple.  Consider a scatterplot or a chloropleth and think about how one might construct templates with flexible x/y or color scales.  Creating separate notebooks is tedious.  Creating cells like "x_scale = d3.scale*..." or "color_scale = d3...." with sensible defaults and allowing customization through robservable seems way better.

Thank you again for writing this excellent package.  My team will be incorporating Observable -> robservable into our Shiny application workflows soon, so you may not have heard the last of me!

**Comment by juba** [comment]
Yes, you're definitely right, this would be useful and we should implement it. I think we just need to find the best way to pass this kind of inputs to `robservable`, either as a specific `input` entry or as a different parameter (`input_js` maybe ?).

**Comment by juba** [comment]
So I just pushed a more complete implementation of cell updating with JavaScript code. It should now support the whole Observable runtime API such as described here :

https://github.com/observablehq/runtime#variable_define

To keep things clean for the most frequent use cases, I added a new `input_js` argument to `robservable`, to be used when a cell must be updated with something else than a "fixed" value. I put some explanations in the introduction vignette : 

https://juba.github.io/robservable/articles/introduction.html#updating-a-cell-javascript-code

Let me know if you think the implementation could be better, or if you come across any bug.
